### PR TITLE
Make networkID as an indication of pod update

### PR DESCRIPTION
**Pull request by hzxuzhonghu**
**Please provide a description of this PR:**

And also rename podInfos

### CODE REVIEW

File: pilot/pkg/model/push_context.go [hunk 1]
```diff
@@ -2164,8 +2164,8 @@ func (ps *PushContext) ServiceAccounts(hostname host.Name, namespace string, por
 	}]
 }
 
-func (ps *PushContext) SupportsTunnel(ip string) bool {
-	infos, _ := ps.ambientIndex.PodInformation(sets.New(types.NamespacedName{Name: ip}))
+func (ps *PushContext) SupportsTunnel(ip string, clusterID string) bool {
+	infos, _ := ps.ambientIndex.WorkloadInfos(sets.New(types.NamespacedName{Name: ip, Namespace: clusterID}))
```
Comment (ymesika) [inline pilot/pkg/model/push_context.go:2168]: Not sure I understand why you are assigning the Cluster ID to the namespace fields throughout this PR.
Can you please explain?

**Review by howardjohn — CHANGES_REQUESTED** [review]
Cluster is not the namespace for IP address, network is. So this seems like something that should be part of multi-network support, which is a lot bigger effort thanthis PR.

Can you give more context?

**Comment by hzxuzhonghu** [comment]
Network is not very right, different clusters can have same pod CIDR within same network. SO the pod IP can be duplicated in same network

File: pilot/pkg/model/push_context.go (hunk 1 above)
Comment (hzxuzhonghu) [reply pilot/pkg/model/push_context.go:2168]: To support multi-cluster pod info get

**Comment by howardjohn** [comment]
I think the definition of an Istio network is non-ovelapping IP. otherwise
it would be completely broke today , no?

**Comment by hzxuzhonghu** [comment]
I am not sure we have any docs saying that. But using cluster to distinguish pod have no sideefect to me

**Comment by howardjohn** [comment]
Ok we should update the docs then. Do you want to submit a PR to istio.io?

**Comment by hzxuzhonghu** [comment]
Yes, i can do. That's not related to this. Think about searching pod by cluster vs network, as each k8s registry has its own clusterID attribute recorded. But we can only cal the network based on network configuration

**Comment by hzxuzhonghu** [comment]
@ramaraochavali wish you to have a look, comparing using clusterID vs networkID

File: pilot/pkg/serviceregistry/kube/controller/ambientindex.go [hunk 2]
```diff
@@ -809,22 +808,25 @@ func (a *AmbientIndex) handleService(obj any, isDelete bool, c *Controller) map[
...
-func (c *Controller) PodInformation(addresses sets.Set[types.NamespacedName]) ([]*model.WorkloadInfo, []string) {
+func (c *Controller) WorkloadInfos(addresses sets.Set[types.NamespacedName]) ([]*model.WorkloadInfo, []string) {
 	if len(addresses) == 0 {
 		// Full update
 		return c.ambientIndex.All(), nil
 	}
 	var wls []*model.WorkloadInfo
 	var removed []string
 	for p := range addresses {
+		if p.Namespace != "" && p.Namespace != c.Cluster().String() {
```
Comment (hzxuzhonghu) [inline pilot/pkg/serviceregistry/kube/controller/ambientindex.go:827]: Using clusterID we can check directly

But using networkID, we need to call the function to calculate per address

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:827]: Calculate what?

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:827]:
```
// return the mesh network for the endpoint IP. Empty string if not found.
func (b *EndpointBuilder) endpointNetwork(endpointIP string) network.ID {
	// If we're building the endpoint based on proxy meta, prefer the injected ISTIO_META_NETWORK value.
	if b.metaNetwork != "" {
		return b.metaNetwork
	}

	return b.controller.Network(endpointIP, b.labels)
}
```

Comment (howardjohn) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go:827]: That calculation is pretty cheap, its just a few conditional lookups. If looking up a network is slow we should optimize that, not break the model.

Addresses are scoped by network, not cluster.

I recommend we bundle making Address network-aware with making it WorkloadEntry (https://github.com/istio/istio/pull/43799/files) so we don't have overlapping/wasted work

Comment (hzxuzhonghu) [reply pilot/pkg/serviceregistry/kube/controller/ambientindex.go]: done

**Review by howardjohn — COMMENTED** [review]
Hey sorry for the holdup on this - some folsk on the ambient call mentioned they wanted to review this which is what the delay was. I think this was @EItanya and @GregHanson maybe, forgot who exactly

**Comment by EItanya** [comment]
Thanks for pointing this out @howardjohn

**Comment by linsun** [comment]
> **Please provide a description of this PR:**
> 
> And also rename podInfos

@hzxuzhonghu could you provide better desp in the future PRs? :) Thanks!

Quick q, does this impact anything related to workload proto to ztunnel? 

+ @kdorosh for review

**Review by kdorosh — CHANGES_REQUESTED** [review]
I think we will also need to update the Lookup itself https://github.com/istio/istio/blob/f7a3f2a28c34ee4fcd3de0b39648a3987142c0bf/pilot/pkg/serviceregistry/kube/controller/ambientindex.go#L65-L66 and ztunnel as a client to include network in the request

**Pull request closed without merging**